### PR TITLE
Add rule to leverage automatic tupling of function parameters in Dotty

### PR DESCRIPTION
**Pull request by marcelocenerine**
http://dotty.epfl.ch/docs/reference/auto-parameter-tupling.html

@olafurpg, is this something you would like to have in scalafix at this point?

### CODE REVIEW

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/DottyAutoTuplingFunctionArgs.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      case Case(Pat.Tuple(args), None, _) :: Nil
+          if args.forall(isOptionallyTypedVarOrWildcard) =>
+        true
+      case _ => false
+    }
+
+  private def isOptionallyTypedVarOrWildcard(pat: Pat): Boolean =
+    pat.is[Pat.Var] || pat.is[Pat.Typed] || pat.is[Pat.Wildcard]
+
+  private def initSpaces(tokens: Seq[Token]) = tokens.takeWhile(_.is[Space])
```
Comment (marcelocenerine) [inline scalafix-core/shared/src/main/scala/scalafix/internal/rule/DottyAutoTuplingFunctionArgs.scala]: I just found myself writing this again 😜

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/DottyAutoTuplingFunctionArgs.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    "Rewrite that removes pattern-matching decomposition if function arguments can be automatically tupled in Dotty."
+
+  override def fix(ctx: RuleCtx): Patch =
+    ctx.tree.collect {
+      case Term.Apply((_, List(pf: Term.PartialFunction)))
+          if canBeAutoTupled(pf) =>
+        val pfTokens = pf.tokens
+
+        (for {
+          leftBrace <- pfTokens.headOption
```
Comment (olafurpg) [inline scalafix-core/shared/src/main/scala/scalafix/internal/rule/DottyAutoTuplingFunctionArgs.scala]: Guard against `.is[LeftBrace/RightBrace]`?

Comment (marcelocenerine) [reply scalafix-core/shared/src/main/scala/scalafix/internal/rule/DottyAutoTuplingFunctionArgs.scala]: I think the curly braces here are mandatory (http://www.lihaoyi.com/post/WartsoftheScalaProgrammingLanguage.html#needing-curliescase-for-destructuring-anonymous-functions). I put the guards anyway to make it explicit  👍

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/DottyAutoTuplingFunctionArgs.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+          rightBrace <- pfTokens.lastOption
+          kwCase <- pfTokens.find(_.is[KwCase])
+          singleLine = pf.pos.startLine == pf.pos.endLine
+        } yield {
+          val bracesReplacedWithParentheses =
+            if (singleLine) {
```
[the remark this answers is not in the document]
Comment (marcelocenerine) [reply scalafix-core/shared/src/main/scala/scalafix/internal/rule/DottyAutoTuplingFunctionArgs.scala]: done

File: scalafix-tests/input/src/main/scala/test/DottyAutoTuplingFunctionArgs.scala [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/*
+rules = DottyAutoTuplingFunctionArgs
+ */
+package test
+
+object DottyAutoTuplingFunctionArgs {
+  List((1, 2), (3, 4)).map { case (a, b) => a + b }
```
[the remark this answers is not in the document]
Comment (marcelocenerine) [reply scalafix-tests/input/src/main/scala/test/DottyAutoTuplingFunctionArgs.scala]: oh, you found a way to break it. Nice catch! Your suggestion https://github.com/scalacenter/scalafix/pull/513/files#r157722339 addresses this.

**Comment by allanrenucci** [comment]
> I also ran it on dotty-compiler-bootstrapped but couldn't test the compilation since it depends on dotty-compiler/compile with scala 2.x

Can you run it on `dotty-compiler` instead and test the compilation by setting the `scalaVersion` value to `0.5.0-RC1`?

**Comment by smarter** [comment]
> I also ran it on dotty-compiler-bootstrapped but couldn't test the compilation since it depends on dotty-compiler/compile with scala 2.x

You can bootstrap from an existing published bootstrap. First publish a working bootstrap:
```scala
> dotty-bootstrapped/publishLocal
```
Then use it to compile the bootstrap:
```scala
> set bootstrapFromPublishedJars in ThisBuild := true
```

**Comment by allanrenucci** [comment]
I have some examples in mind that may be tricky to rewrite:
```scala
scala> class A 
// defined class A
scala> class B extends A 
// defined class B
scala> def foo(xs: List[(A, A)]) = xs.map((x, y) => 1) 
def foo(xs: List[(A, A)]): List[Int]
scala> def foo(xs: List[(A, A)]) = xs.map((x: A, y: A) => 1) 
def foo(xs: List[(A, A)]): List[Int]
scala> def foo(xs: List[(A, A)]) = xs.map((x: B, y: A) => 1) 
1 |def foo(xs: List[(A, A)]) = xs.map((x: B, y: A) => 1)
  |                                                    ^
  |                                                 found:    (B, A) => Int
  |                                                 required: ((A, A)) => Int
  |                                                 
scala> def foo(xs: List[(A, A)]) = xs.map { case (x: B, y: A) => 1 }  // Note this one cannot be rewritten
def foo(xs: List[(A, A)]): List[Int]
scala> def foo(xs: List[(A, A)]) = xs.map((x: Any, y: A) => 1) 
def foo(xs: List[(A, A)]): List[Int]
```

**Comment by marcelocenerine** [comment]
Interesting example... I don't think there is enough type information in scalameta (nor if it is intended to) to allow the rule to evaluate the type relationship between the partial function parameters and the tuple elements so it can be 100% sure that the pattern matching is not needed... 

To be on the safe side, maybe the rule could only rewrite when the partial function parameters are explicitly typed and the types exactly match the types of the corresponding elements in the tuple. This might be doable....

**Comment by olafurpg** [comment]
> Interesting example... I don't think there is enough type information in scalameta (nor if it is intended to in the future) to allow the rule to evaluate the type relationship between the partial function parameters and the tuple elements so it can be 100% sure that the pattern matching is not needed...

We actually have the full signature with attributed types of the invoked symbol. However, there is no `=:=` that handles de-aliasing etc. so I think it's best for now to edge on the safe side and leave `pat.is[Pat.Typed]` out of the cases that get rewritten.

**Comment by marcelocenerine** [comment]
>We actually have the full signature with attributed types of the invoked symbol. 

sorry, I meant that it may not be possible to determine if `B <: A` (I could be wrong)

>I think it's best for now to edge on the safe side and leave pat.is[Pat.Typed] out of the cases that get rewritten.

Done

File: website/src/main/tut/docs/rules/DottyAutoTuplingFunctionArgs.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+Removes pattern-matching decomposition if function arguments can be automatically tupled. See [http://dotty.epfl.ch/docs/reference/auto-parameter-tupling.html](http://dotty.epfl.ch/docs/reference/auto-parameter-tupling.html)
+
+```scala
+// before
+xs.map { case (x, y) => x + y }
+// after
+xs.map((x, y) => x + y)
+```
+
+Auto-tupling is a language feature only supported in Dotty. Therefore, the code produced by this rewrite does not compile with Scala 2.x.
```
Comment (olafurpg) [inline website/src/main/tut/docs/rules/DottyAutoTuplingFunctionArgs.md]: 👍

**Comment by olafurpg** [comment]
Here is a diff from running the rewrite on the dotty repo including submodules
https://gist.github.com/olafurpg/d6a393d16487b50a39d8a09a3a825145
@allanrenucci I got
```
[error] (dotty-compiler/compile:compileIncremental) java.lang.NoClassDefFoundError: dotty/tools/dotc/interfaces/ReporterResult
```
when running `++0.5.0-RC1`.  I also tried `set bootstrapFromPublishedJars in ThisBuild := true` after publishLocal but that still compiled dotty-compiler with Scala 2.x. I tested this on scalatest and everything still compiled after.

The diff  in apache/spark is 11k LOC *_* https://gist.github.com/olafurpg/1b52f2b3d070dc8424fd41aa385b9873

**Comment by allanrenucci** [comment]
Actually, I don't think this rewrite can be syntactic:
```scala
scala> def foo(xs: List[Any]) = xs.map { case (x, y) => 1 } 
def foo(xs: List[Any]): List[Int]
scala> def foo(xs: List[Any]) = xs.map((x, y) => 1 ) 
1 |def foo(xs: List[Any]) = xs.map((x, y) => 1 )
  |                                ^^^^^^^^^^^
  |                                wrong number of parameters, expected: 1
```

**Comment by marcelocenerine** [comment]
😞 I'll take a look at #521

**Comment by olafurpg** [comment]
Good catch! @allanrenucci I opened https://github.com/scalacenter/scalafix/issues/521 with details to followup on that. Am I right in thinking that this case only happens for `Any`/`AnyRef` since they are supertypes of `TupleN`?

**Comment by olafurpg** [comment]
@marcelocenerine No worries, getting rewrites 100% correct is really hard! I also think there is value in having a syntactic rewrite that runs fast without compilation, and manually fixing corner cases where you are doing silly stuff like deconstructing `Any` into `(Any, Any)`. Most important bit is to set the expectations correctly in the docs.

**Comment by smarter** [comment]
> I also think there is value in having a syntactic rewrite that runs fast without compilation, and manually fixing corner cases

That sounds exactly like what I don't want in an automatic rewrite :P. It can take as long as it wants as, that's fine as long as I can trust the result.

**Comment by allanrenucci** [comment]
@olafurpg I think it is more complicated than that. You are basically doing exhaustivity checking here. So off the top of my head: `ProductN`, type variables and am sure I am missing many other corner cases. Maybe @liufengyun can comment on this,.

Also, you don't want to rewrite something that expect a partial function for example:
```scala
def foo(f: PartialFunction[(Int, Int), Int]): Nothing
foo { case (x, y) => 1 } // should't be rewritten.
```

**Comment by olafurpg** [comment]
> That sounds exactly like what I don't want in an automatic rewrite :P. It can take as long as it wants, that's fine as long as I can trust the result.

If I have the option between slow correct rewrite and fast but slightly buggy rewrite then I definitely agree with you and pick the slow correct one 😄 But our current choice is unfortunately between no rewrite and fast but slight buggy rewrite (failing in cases where you deconstruct into `(Any, Any)`) so I personally prefer the latter setting expectations correctly in the docs #522 . Migration tools like codemod https://github.com/facebook/codemod are widely used despite producing non-perfect rewrites.

**Comment by marcelocenerine** [comment]
There are a number of missed corner cases pointed out above and in #521. @olafurpg If you agree I'd suggest to revert my commit for now until the rule can handle those scenarios correctly and produce reliable results

**Comment by olafurpg** [comment]
Sure we can revert the commit @marcelocenerine

**Pull request merged**